test/sqllogictest/cast.slt

```diff
@@ -115,5 +123,5 @@ SELECT date('2020-01-01'), date('2020-01-01'::timestamp), date('2020-01-01'::tim
 query error invalid input syntax
```

Comment:
Ah, this error is insufficiently specified.

We need something akin to this diff:

```diff
diff --git a/src/pgrepr/src/oid.rs b/src/pgrepr/src/oid.rs
index e492c5fb9..b717c4f18 100644
--- a/src/pgrepr/src/oid.rs
+++ b/src/pgrepr/src/oid.rs
@@ -203,3 +203,4 @@ pub const FUNC_MZTIMESTAMP_GTE_MZTIMESTAMP_OID: u32 = 16_559;
 pub const FUNC_MZ_NOW_OID: u32 = 16_560;
 pub const FUNC_MAX_MZTIMESTAMP_OID: u32 = 16_561;
 pub const FUNC_MIN_MZTIMESTAMP_OID: u32 = 16_562;
+pub const FUNC_DATE_FROM_TEXT: u32 = 16_563;
diff --git a/src/sql/src/func.rs b/src/sql/src/func.rs
index 13ce08e42..59b5bac7c 100644
--- a/src/sql/src/func.rs
+++ b/src/sql/src/func.rs
@@ -1874,6 +1874,7 @@ pub static PG_CATALOG_BUILTINS: Lazy<HashMap<&'static str, Func>> = Lazy::new(||
             params!(Int32) => UnaryFunc::Chr(func::Chr), 1621;
         },
         "date" => Scalar {
+            params!(String) => UnaryFunc::CastStringToDate(func::CastStringToDate), oid::FUNC_DATE_FROM_TEXT;
             params!(Timestamp) => UnaryFunc::CastTimestampToDate(func::CastTimestampToDate), 2029;
             params!(TimestampTz) => UnaryFunc::CastTimestampTzToDate(func::CastTimestampTzToDate), 1178;
         },
```

In this case we need to allocate a new oid because PG doesn't handle this case as a function; they must have some special case that sniffs out typename-looking functions with a parameter of an unknown type (which means it looks like a string) and instead does a parse.

This then lets us return the correct error here, which is:

```
invalid input syntax for type date
```

(without this diff it tries to cast the parameter to a ts)

Think this is important to do because this implementation is likely to be the template everyone follows for adding these functions in the future and better to propagate this now.

Reply:
Ah that's a great suggestion